sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -493,9 +510,23 @@ case class DataSource(
         dataSource.createRelation(
           sparkSession.sqlContext, mode, caseInsensitiveOptions, Dataset.ofRows(sparkSession, data))
       case format: FileFormat =>
-        sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd
+        val cmd = planForWritingFileFormat(format, mode, data)
+        val resolvedPartCols = cmd.partitionColumns.map { col =>
+          // The partition columns created in `planForWritingFileFormat` should always be
+          // `UnresolvedAttribute` with a single name part.
+          assert(col.isInstanceOf[UnresolvedAttribute])
+          val unresolved = col.asInstanceOf[UnresolvedAttribute]
+          assert(unresolved.nameParts.length == 1)
+          val name = unresolved.nameParts.head
+          outputColumns.find(a => equality(a.name, name)).getOrElse {
+            throw new AnalysisException(
+              s"Unable to resolve $name given [${data.output.map(_.name).mkString(", ")}]")
+          }
+        }
+        val resolved = cmd.copy(partitionColumns = resolvedPartCols, outputColumns = outputColumns)
```

Comment:
Is this ad-hoc column resolution just to ensure the names have the correct case after it is possibly dropped by the optimizer? Why does the command need to report these and where are they used?

Reply:
The previous code calls `sparkSession.sessionState.executePlan` to analze/optimize/plan/exeucte this temporary `InsertIntoHadoopFsRelationCommand`, which is pretty hacky because at this moment, we are executing CTAS and we already have the final physical plan. Here we manually analyze the `InsertIntoHadoopFsRelationCommand` so that we only reuse the physical part here.

Reply:
I get the point of passing the physical plan, and I think that's a good idea. What I don't understand is why the command doesn't match the physical plan that is passed in. Is that physical plan based on a different logical plan? I would expect that the physical plan is created once and passed into run, but that it was created from the logical plan that is also passed in.

Reply:
The given physical plan has been registered to the UI, and we can collect its metrics if we execute it. However if we run `sparkSession.sessionState.executePlan`, we get a new physical plan which semantically equals to the given physical plan but not the same object. This new physical plan is not registered to UI so we can't show metrics correctly in the UI.

Reply:
I don't think you're answering the question that I'm asking. I understand why the physical plan is passed in.

Why does the physical plan not match the command that is produced, or why doesn't the command here match the physical plan? I don't see why executePlan would produce something different than what is passed in.

Reply:
> Why does the physical plan not match the command that is produced

It matches! The only problem is, they are 2 different JVM objects. The UI keeps the physical plan object and displays them. An alternative solution is to swap the new physical plan into the UI part, but that's hard to do with the current UI framework.

If we run `sparkSession.sessionState.executePlan(planForWritingFileFormat(format, mode, data)).toRdd`, we are executing the new physical plan, so no metrics will be reported to the passed-in physical plan and shown in the UI.

Reply:
If it matches, then why is there column resolution happening?

Reply:
`planForWritingFileFormat` returns an **unresolve** `InsertIntoHadoopFsRelationCommand`. The previous code runs `sparkSession.sessionState.executePlan` to analze/.../exeucte the unresolved `InsertIntoHadoopFsRelationCommand`, which convert logical plan to physical plan and execute it.

Now, I wanna run the given physical plan. Even though `sparkSession.sessionState.executePlan` can produce an exactly same physical plan, they are different objects and UI is not happy with it, so I choose to manually resolve the unresolved `InsertIntoHadoopFsRelationCommand` here, and explicitly ask it to run my given physical plan.

Reply:
Ok, I think I get it now. Thanks for explaining.

Why not pass the `QueryExecution` so you have access to the resolved plan without copying resolution rules here? I'm just curious here, I get that this is intended as a quick fix for the release, so don't let my comments block you.

Reply:
Because the CTAS node itself doesn't have a `QueryExecution`. It only has the physical plan that was produced by the planner. `QueryExecution` only exists in the places that drive the  analze/.../exeucte, e.g. `Dataset`.